src/core/backbone_router/bbr_manager.cpp

```diff
@@ -90,18 +111,63 @@ void Manager::HandleMulticastListenerRegistration(const Coap::Message &aMessage,
     // TODO: (MLR) send configured MLR response for Reference Device
     // TODO: (MLR) handle Commissioner Session TLV
     // TODO: (MLR) handle Timeout TLV
-    // TODO: (MLR) handle addresses, updated listener groups, send BMLR.req
+
+    VerifyOrExit(ThreadTlv::FindTlvValueOffset(aMessage, IPv6AddressesTlv::kIPv6Addresses, addressesOffset,
+                                               addressesLength) == OT_ERROR_NONE,
+                 error = OT_ERROR_PARSE);
+    VerifyOrExit(addressesLength % sizeof(Ip6::Address) == 0, error = OT_ERROR_PARSE);
+    VerifyOrExit(addressesLength / sizeof(Ip6::Address) <= kIPv6AddressesNumMax, error = OT_ERROR_PARSE);
+
+    SuccessOrExit(error = Get<BackboneRouter::Leader>().GetConfig(config));
+    expireTime = TimerMilli::GetNow() + TimeMilli::SecToMsec(config.mMlrTimeout);
+
+    for (uint16_t offset = 0; offset < addressesLength; offset += sizeof(Ip6::Address))
+    {
+        bool failed = true;
+
+        IgnoreReturnValue(aMessage.Read(addressesOffset + offset, sizeof(Ip6::Address), &address) ==
+                          sizeof(Ip6::Address));
+
+        switch (mMulticastListenersTable.Add(address, expireTime))
+        {
+        case OT_ERROR_NONE:
+            failed = false;
+            break;
+        case OT_ERROR_INVALID_ARGS:
+            if (status == ThreadStatusTlv::kMlrSuccess)
+            {
+                status = ThreadStatusTlv::kMlrInvalid;
+            }
+            break;
+        case OT_ERROR_NO_BUFS:
```

Comment:
When `NO_BUFS` happens, may skip further `mMulticastListenersTable.Add` and just take following addresses as failed?

Reply:
I prefer the current code:

- To implement what you have suggested, we need extra code to stop the loop and append left addresses (unprocessed) to `failedAddresses`, which will increase code size
- The current code can already do that (even though with a little more CPU cycles)
- If there is a valid address that was already registered, the current code can correctly handle it without adding it to the `failedAddresses`, even if the table is running out of memory space.